UUID validators with custom string object

Hi,

I use the Pydantic version `2.6.0`.

When trying to validate an UUID from a customize string (the class _CustomizeStr in the code bellow), the validation fails (i.e.  `v.validate_python(_CustomizeStr('1fa79205-f747-465b-bf7a-cabc8ce60644'))`) with the following error:

```
>       output = v.validate_python(_CustomizeStr('1fa79205-f747-465b-bf7a-cabc8ce60644'))
E       pydantic_core._pydantic_core.ValidationError: 1 validation error for uuid
E         UUID input should be a string, bytes or UUID object [type=uuid_type, input_value=1fa79205-f747-465b-bf7a-cabc8ce60644, input_type=test_uuid_with_custom_string.<locals>._CustomizeStr]
E           For further information visit https://errors.pydantic.dev/latest/v/uuid_type
```

Whereas with Python UUID class, it is possible to create an UUID from a customize string i.e. `UUID(_CustomizeStr('1fa79205-f747-465b-bf7a-cabc8ce60644'))`. So I would expect the same behaviour with Pydantic.

[The test to highlight the issue I encounter](https://github.com/fafanoulele/pydantic-core/commit/55efcd00e2e81e4a2d021b1f1e4b53c55d03310d):
```python
def test_uuid_with_custom_string():
    class _CustomizeStr(str):
        def __init__(self, original_str: str):
            super(_CustomizeStr, self).__init__()
            self.__class__.__name__ = 'str'
            self._original_str = original_str

        def __str__(self) -> str:
            return self._original_str

        def __repr__(self) -> str:
            return self._original_str

    assert UUID('1fa79205-f747-465b-bf7a-cabc8ce60644') is not None
    assert UUID(_CustomizeStr('1fa79205-f747-465b-bf7a-cabc8ce60644')) is not None

    v = SchemaValidator({'type': 'uuid'})
    # The following line fails
    output = v.validate_python(_CustomizeStr('1fa79205-f747-465b-bf7a-cabc8ce60644'))
    assert output == UUID('1fa79205-f747-465b-bf7a-cabc8ce60644')
    assert isinstance(output, UUID)
```

To be honest, I am not sure it is valid to consider that as an issue. On my side, I've workaround the issue, but I wanted to bring up that to have your opinion. Feel free to close it if it is not a valid issue.

--- a/src/validators/uuid.rs
+++ b/src/validators/uuid.rs
@@ -12,6 +12,7 @@ use crate::errors::{ErrorType, ErrorTypeDefaults, ValError, ValResult};
 use crate::input::input_as_python_instance;
 use crate::input::Input;
 use crate::input::InputType;
+use crate::input::ValidationMatch;
 use crate::tools::SchemaDict;
 
 use super::model::create_class;
@@ -152,7 +153,7 @@ impl Validator for UuidValidator {
 
 impl UuidValidator {
     fn get_uuid<'py>(&self, input: &(impl Input<'py> + ?Sized)) -> ValResult<Uuid> {
-        let uuid = match input.exact_str().ok() {
+        let uuid = match input.validate_str(true, false).ok().map(ValidationMatch::into_inner) {
             Some(either_string) => {
                 let cow = either_string.as_cow()?;
                 let uuid_str = cow.as_ref();

--- a/tests/validators/test_uuid.py
+++ b/tests/validators/test_uuid.py
@@ -9,6 +9,9 @@ from pydantic_core import SchemaValidator, ValidationError, core_schema
 from ..conftest import Err, PyAndJson
 
 
+class MyStr(str): ...
+
+
 @pytest.mark.parametrize(
     'input_value,expected',
     [
@@ -25,6 +28,7 @@ from ..conftest import Err, PyAndJson
         ('c0a8f9a8-aa5e-482b-a067-9cb3a51f5c11', UUID('c0a8f9a8-aa5e-482b-a067-9cb3a51f5c11')),
         ('00000000-8000-4000-8000-000000000000', UUID('00000000-8000-4000-8000-000000000000')),
         ('00000000-0000-4000-0000-000000000000', UUID('00000000-0000-4000-0000-000000000000')),
+        (MyStr('00000000-0000-4000-0000-000000000000'), UUID('00000000-0000-4000-0000-000000000000')),
         (b'\x12\x34\x56\x78' * 4, UUID('12345678-1234-5678-1234-567812345678')),
         (b'\x00\x00\x00\x00' * 4, UUID('00000000-0000-0000-0000-000000000000')),
         (b'ebcdab58-6eb8-46fb-a190-d07a33e9eac8', UUID('ebcdab58-6eb8-46fb-a190-d07a33e9eac8')),

